src/backend/distributed/transaction/backend_data.c

```diff
@@ -1278,3 +1290,202 @@ DecrementExternalClientBackendCounter(void)
 {
 	pg_atomic_sub_fetch_u32(&backendManagementShmemData->externalClientBackendCounter, 1);
 }
+
+
+/*
+ * GetEpochTimeMs returns the epoch value in milliseconds.
+ */
+uint64
+GetEpochTimeMs(void)
+{
+	struct timeval tp;
+
+	gettimeofday(&tp, NULL);
+
+	uint64 result = (uint64) (tp.tv_sec) * 1000;
+	result = result + (uint64) (tp.tv_usec) / 1000;
+	return result;
+}
+
+
+/*
+ * get_cluster_clock() is an UDF that returns a monotonically increasing logical
+ * clock. Clock guarantees to never go back in value after restarts, and makes
+ * best attempt to keep the value close to unix epoch time in milliseconds.
+ */
+Datum
+get_cluster_clock(PG_FUNCTION_ARGS)
+{
+	uint64 nextClusterClockValue = GetClusterClock();
```

Comment:
should call CheckCitusVersion(ERROR); first

Reply:
Makes sense, done.